source/core/DualTypeVisitor.cpp

```diff
@@ -29,97 +29,99 @@ namespace Vireo
+
+        if (variantX->IsUninitializedVariant() && variantY->IsUninitializedVariant())
+            return true;
+
+        if (variantX->IsUninitializedVariant() != variantY->IsUninitializedVariant())
             return false;
-        } else {
-            success = TypesAreCompatible(variantUnderlyingTypeX,
-                variantUnderlyingTypeX->Begin(kPARead),
-                variantUnderlyingTypeY,
-                variantUnderlyingTypeY->Begin(kPARead),
-                operation);
+
+        if (variantX->HasData()  != variantY->HasData())
+            return false;
+
+        if (variantX->HasMap() != variantY->HasMap())
+            return false;
+
+        if (variantX->HasMap() && variantY->HasMap() &&
+            variantX->GetMap_size() != variantY->GetMap_size()) {
+                return false;
         }
-        if (success) {
-            // compare attributes
-            if (!variantTypeX->_attributeMap && !variantTypeY->_attributeMap) {
-                return true;
-            } else if (variantTypeX->_attributeMap && variantTypeY->_attributeMap) {
-                if (variantTypeX->_attributeMap->size() != variantTypeY->_attributeMap->size()) {
-                    return false;
-                }
-                for (const auto attributePairInX : *variantTypeX->_attributeMap) {
-                    StringRef const attributeNameStrInX = attributePairInX.first;
-                    VariantTypeRef attributeValueInX = attributePairInX.second;
-                    auto attributePairInY = variantTypeY->_attributeMap->find(attributeNameStrInX);
-                    if (attributePairInY != variantTypeY->_attributeMap->end()) {
-                        VariantTypeRef attributeValueInY = attributePairInY->second;
-                        if (!TypesAreCompatible(attributeValueInX,
-                            attributeValueInX->Begin(kPARead),
-                            attributeValueInY,
-                            attributeValueInY->Begin(kPARead),
-                            operation)) {
-                            return false;
-                        }
-                    } else {
-                        return false;
-                    }
-                }
-            } else {
+
+        // All the inexpensive checks are done first (above)
+        // Do the expensive checks below
+
+        TypeRef underlyingTypeX = variantX->GetInnerType();
+        TypeRef underlyingTypeY = variantY->GetInnerType();
+
+        if (underlyingTypeX != nullptr && underlyingTypeY != nullptr) {
+            VIREO_ASSERT(variantX->HasData() && variantY->HasData());
+            if (!TypesAreCompatible(underlyingTypeX,
+                variantX->GetInnerData(),
+                underlyingTypeY,
+                variantY->GetInnerData(),
+                operation)) {
                 return false;
             }
         }
-        return success;
+
+        if (variantX->HasMap() && variantY->HasMap()) {
```

Comment:
The previous implementation was comparing attribute-names by querying the name from one map against the other, we don't seem to do that anymore. Is that correct?

Reply:
Well,  just noticed that we compare names as part of the Apply operation. However, by comparing attribute values in the order in which they appear, we could likely be comparing TypeAreCompatible on attribute values that are not associated with the same attribute name. But I think that is fine because it would help us filter out a subset of incompatible maps, and then verify them again during Apply. So I think it is fine this way.

Reply:
Yeah, tried to follow: "check for types in *Compatible method and look at values in Apply* method" as type checks are supposed to be faster than value checks.